posix: fix one-time timer for SIGEV_SIGNAL

n-gillet opened the pull request:
For SIGEV_SIGNAL, the function zephyr_timer_wrapper() is the handler between kernel and posix layer. Here, for one-time timer, reload is equal to 0 and function returns. As a consequence, handler function was never called.

Add test one one-time SIGEV_SIGNAL timer
- Set a one shot timer to expiry in 1s.
- After 5s, check that the timer handler has been called only once. Without the fix, it was never called.

cfriedt commented on tests/posix/common/src/timer.c:
```diff
@@ -104,7 +104,7 @@ ZTEST(timer, test_CLOCK_MONOTONIC__SIGEV_THREAD)
 ZTEST(timer, test_timer_overrun)
 {
-	struct sigevent sig = { 0 };
+	struct sigevent sig = {0};
```
My preference would be to keep changes that are exclusively formatting in a separate commit, if they are necessary.

n-gillet replied:
commit restored (not squashed)

ycsin replied:
This change (the 2nd commit) should have a commit message

cfriedt replied:
This change is still not strictly unnecessary. If you feel we need it, then please add it as a 3rd commit, or simply undo this change (you night need to disable auto-formatting).

cfriedt reviewed (changes requested):
Please see comments.

n-gillet commented on tests/posix/common/src/timer.c:
```diff
@@ -122,6 +122,29 @@ ZTEST(timer, test_timer_overrun)
+	value.it_value.tv_sec = 1;
+	value.it_value.tv_nsec = 0;
+	zassert_ok(timer_settime(timerid, 0, &value, NULL));
+	k_sleep(K_MSEC(5000));
```
test updated as suggested

ycsin replied:
This change should be squashed with the first commit

n-gillet replied:
squashed

ycsin commented:
@n-gillet it is very close now, please take a look at the comments, should be easy to fix!

cfriedt commented:
@n-gillet - there seems to be some miscommunication about what commits we expect. Please create

1. code commit (change to `lib/posix/options/timer.c`)
2. test commit (added `test_one_shot__SIGEV_SIGNAL` testcase, increasing coverage, which includes newly added code)
3. formatting commit (to change `{ 0 }` to `{0}` in old code)

n-gillet commented:
> @n-gillet - there seems to be some miscommunication about what commits we expect. Please create
> 
>     1. code commit (change to `lib/posix/options/timer.c`)
> 
>     2. test commit (added `test_one_shot__SIGEV_SIGNAL` testcase, increasing coverage, which includes newly added code)
> 
>     3. formatting commit (to change `{ 0 }` to `{0}` in old code)

ok. But it will not be as suggested in https://github.com/zephyrproject-rtos/zephyr/pull/80687#discussion_r1827730273
thanks for the clarification

Commit:
posix: fix one-time timer for SIGEV_SIGNAL

For SIGEV_SIGNAL, the function zephyr_timer_wrapper() is the handler
between kernel and posix layer.
Here, for one-time timer, reload is equal to 0 and function returns.
As a consequence, handler function was never called.

Signed-off-by: Noemie Gillet <ngillet@sequans.com>

Commit:
test: posix: add test_one_shot__SIGEV_SIGNAL testcase

increase test coverage for newly added code

Signed-off-by: Noemie Gillet <ngillet@sequans.com>

Commit:
test: posix: timer: fix format

change { 0 } to {0}

Signed-off-by: Noemie Gillet <ngillet@sequans.com>

cfriedt commented on tests/posix/common/src/timer.c:
```diff
@@ -122,6 +122,29 @@ ZTEST(timer, test_timer_overrun)
 
+ZTEST(timer, test_one_shot__SIGEV_SIGNAL)
+{
+	struct sigevent sig = {0};
```
This line of code should only be added in the second commit. There is no need to change it as well in the third commit.

n-gillet replied:
Sorry for this. Commit for formatting existing code should only change existing code now.
Concerning https://github.com/zephyrproject-rtos/zephyr/pull/80687#pullrequestreview-2413734406, not sure to get all what you mean by 

> edit the one line so that it reads as {0}

cfriedt reviewed (changes requested):
Very close - only one small change request from me.

You should be able to use

```shell
git rebase -i HEAD~2
# select 'e' for the second commit
# edit the one line so that it reads as {0}
git rebase --continue
```

ycsin commented on tests/posix/common/src/timer.c:
```diff
@@ -104,7 +104,7 @@ ZTEST(timer, test_CLOCK_MONOTONIC__SIGEV_THREAD)
 ZTEST(timer, test_timer_overrun)
 {
-	struct sigevent sig = { 0 };
+	struct sigevent sig = {0};
```
sorry, the title of this commit 7ac39644153742f42b95025d9fd4d93433996267 should be something like `test: posix: timer: fix format` to reflect the 'area' as noted in the contribution guideline

ycsin replied:
just also note that this change might conflict with the relocation #80670

ycsin replied:
> > style: for code style changes
> 
> in the contribution guideline. I though it was for this kind of commit.

I'd use that title if the commit modified more than a few areas. Your commit only affected one file, it is more informative to have a specific area, so that we know what this commit does when scrolling through the git log, without having to click into it.

n-gillet commented:
@ycsin @cfriedt I think I've covered all your comments. Let me know if you want other modifications.

ycsin reviewed (approved):
LGTM

n-gillet commented:
Thanks for your review

dkalowsk commented:
@cfriedt is this needed for 4.0?  I don't see an associate GitHub issue which it'd need at this point.

cfriedt commented:
@dkalowsk it would be a good bug fix to get in ahead of release.

@n-gillet - can you please create a bug report in GitHub to link to this PR?
